components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2IntrospectionAuthenticationHandler.java

```diff
@@ -0,0 +1,156 @@
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+import java.nio.charset.Charset;
+
+/**
+ * OAuth2IntrospectionAuthenticationHandler is for authenticate the request based on consumer key and secret.
+ * canHandle method will confirm whether this request can be handled by this authenticator or not.
+ */
+public class OAuth2IntrospectionAuthenticationHandler extends AuthenticationHandler {
+
+    private static final Log log = LogFactory.getLog(OAuth2IntrospectionAuthenticationHandler.class);
+    private final String INTROSPECTION_URI = "oauth2/introspect";
+    private final String BASIC_AUTH_HEADER = "Basic";
+    private final String CONSUMER_KEY = "consumer-key";
+
+    @Override
+    protected AuthenticationResult doAuthenticate(MessageContext messageContext) throws AuthServerException,
+            AuthenticationFailException, AuthClientException {
+
+        BasicAuthenticationHandler basicAuthHandler = new BasicAuthenticationHandler();
+        AuthenticationResult authenticationResult = basicAuthHandler.doAuthenticate(messageContext);
+
+        if(authenticationResult.getAuthenticationStatus().equals(AuthenticationStatus.SUCCESS)) {
+            return authenticationResult;
+        }
+
+        authenticationResult = new AuthenticationResult(AuthenticationStatus.FAILED);
+        AuthenticationContext authenticationContext = (AuthenticationContext) messageContext;
+        String authorizationHeader = authenticationContext.getAuthenticationRequest().
+                getHeader(HttpHeaders.AUTHORIZATION);
+
+        String[] splitAuthorizationHeader = authorizationHeader.split(" ");
+        if (splitAuthorizationHeader.length == 2) {
+            byte[] decodedAuthHeader = Base64.decodeBase64(splitAuthorizationHeader[1].getBytes());
+            String[] splitCredentials = new String(decodedAuthHeader, Charset.defaultCharset())
+                    .split(Constants.AUTH_HEADER_SEPERATOR, 2);
+            if (splitCredentials.length == 2) {
+                String clientId = splitCredentials[0];
+                String clientSecret = splitCredentials[1];
+
+                try {
+                    boolean isAuthenticated = OAuth2Util.authenticateClient(clientId, clientSecret);
+                    if ( isAuthenticated ) {
+                        authenticationResult.setAuthenticationStatus(AuthenticationStatus.SUCCESS);
+                        if ( log.isDebugEnabled() ) {
+                            log.debug("Basic authentication success for client : "+ clientId + ".");
+                        }
+
+                        OAuthConsumerDAO oAuthConsumerDAO = new OAuthConsumerDAO();
+                        String username = oAuthConsumerDAO.getAuthenticatedUsername(clientId, clientSecret);
+                        if(StringUtils.isNotBlank(username)) {
```

Comment:
What if the "username" null, i.e. app has no owner ? we need to gracefully handle it